### ISSUE TITLE
node js 웹 구성 시 Web 서버와 WAS 서버

### ISSUE THREAD

**Issue by bin1509203** [codingeverybody/codingyahac#1002]
다소 이번 AI강좌와는 거리가 좀 있는 문제이지만... 어디에 물어볼 곳이 없어 질문 드립니다... 양해 부탁드립니다!
본 강좌와 관련된 질문 먼저 답변 주시고 시간 남으시면 의견 부탁드려요..!

# 해결하고자 하는 문제
생활코딩과 구글 검색을 통해 배운 지식으로 웹어플리케이션 개발을 하다보니 "기능 구현" 이외의 문제에서 막히게 되었습니다.
테스트용 웹은 node.js를 활용해 1개의 단일 서버로 구현하였었는데 실제 상용 서비스를 위해서는 보안상의 이유로 구조를 3단계로  해야하더라구요.
**"DMZ 내에 Web서버" + 방화벽 + "WAS 서버" + "DB 서버"**

현재 구현한 코드를 기능 별로 나눠보면 6가지 정도 되는데 이걸 기능에 맞게 Web서버와 WAS서버로 나누는 것인지 등
나누는 기준이 궁금합니다. 프론트/백 모두 node.js로 구현되어 있어서 아파치나 톰캣, php 등은 사용하지 않았습니다.

1) html 부분(pug 이용)
2) css 및 js 부분(부트스트랩 등)
3) node.js main 코드(router, https 모듈)
4) node.js config 코드(express, db 등)
5) node.js sub 코드(routes, 렌더링 및 리다이렉트, 서버 단 기능 구현 부분)
6) 기타 기능 구현을 위한 python 코드

# 환경
Ubuntu Server, node.js 최신 버전, python3.7, mariaDB 최신 버전

# 시도해본 방법
Web 서버 : 1) html 2) css/js 3) node.js main
WAS 서버 : 4) node.js config 5) node.js sub 6) python

**Comment by progh2** [comment]
제 생각에는 Web서버와 WAS서버를 나누는 경우는 보통 스프링이나 JSP 같은 경우에 나누고 node.js의 경우에는 특별히 static 파일들을 아파치 웹서버에서 담당하게 할 것이 아니라면 나누지 않아도 될 것 같습니다. 
(CDN을 쓸 수 있다면 static 파일들을 CDN에 빼도 될 것 같네요)
그리고 서비스 규모가 크지 않다면 저정도로 쪼개지 않아도 될 것 같아요. 서비스 요청이 매우 많아져서 정적 파일들은 아파치에게 맡기고 node.js의 성능을 쫌 더 쥐어 짜고 싶다면 나눠볼만 할 것 같아요~

**Comment by bin1509203** [comment]
회사 보안 규정 상 규모와 상관없이 WEB과 WAS를 나누기를 요구하고 있는데 말씀하신것처럼 
대부분의 엔터프라이즈급 개발은 스프링으로 이루어지는 터라, nodejs에 대한 명확한 가이드가 없네요... ㅠㅠ

구글에서 검색해보면 보통 아파치 + node.js 를 쓰는 경우에는 아파치를 Reverse Proxy로 사용한다고 나와 있던데
static 파일을 아파치에 맡긴다는것이랑 같은 맥락에서의 이야기인가요?

**Comment by progh2** [comment]
네 비슷합니다. 특정 경로로 들어오는 요청은 node.js로 넘겨서 처리하고 일반적인 요청은 아파치가 감당하도록 할 때 말씀대로 설정 넣어서 쓸 수 있어요. 
음.. 저는 학교에서 학생들을 위한 서버를 운영 중인데, 하나의 서버에 여러 도메인을 할당해서 아파치웹서버, jsp, php, node.js, django 등을 제공하고 있는데 이때도 사용하고 있습니다. 

node.js를 전문으로 서비스하는 개발자분의 말씀도 추가로 들어봐야 할 것 같네요.

**Comment by fermat39** [comment]
인프라 구성에 대해서 좀 설명이 필요할 것 같습니다.

웹서버는 말 그대로 기본적인 웹서비스를 위한 서버입니다. 사용자의 접속에 대해서
제일 앞단에서 처리를 하게 되지요. 이 웹서버에서 로드밸린싱을 보통하게 됩니다.
이 웹서버가 1대 일수도 있고 여러대 일 수 있습니다. 실제에서는 서비스 IP로 접속을
하게 되어서 1대로 보일 수 있습니다.

이어서 웹서버가 실제로 비지니스 로직을 처리하는 애플리케이션 서버에 접속하게 됩니다. 
이 애플리케이션 서버는 실제로 프로그래밍 언어, jsp, php, python등을 통해 로직을 처리하게 됩니다.
이 애플리케이션 서버도 여러 개로 운영될 수 있습니다. 여러 개라는 의미는 물리적인 경우도 있고,
같은 서버 안에서 port로 분리되어 운영될 수도 있습니다. 

https://www.tecmint.com/best-open-source-web-servers/ 가 조금이나마 도움이 되셨으면 합니다.
또 이해가 안 가시면 다시 문의주세요~~
